### ISSUE TITLE
Class in markdown image

### ISSUE THREAD

**Issue by lebarde** [russross/blackfriday#364]
There is no convention about adding a CSS class to an image. What is often done is:

- `![my picture title](url){.my-class}`, or
- `![my picture title](url){: class="my-class"}`

or some else notation. Personnally, I like pretty much this second option, which enables to add raw HTML attributes.

Actually the image conventions in markdown are is too small for my needs.

**Comment by shurcooL** [comment]
Since you haven't mentioned it, an existing solution to add a class to an image is to use an HTML `<img>` tag. I.e.:

```HTML
<img class="my-class" src="https://www.example.org/image.png">
```

**Comment by lebarde** [comment]
Absolutely. There is also a CSS trick to select a picture: use `![my picture title](url#fig)` inside Markdown, and then trigger it with that:

``` css
img[src*="#fig"] {
    float: left;
    margin: 30px;
}
```

But I need to add specific classes to get images handled with Bootstrap stylesheet (`.img-fluid`, and so on). I think I am not alone in this case.

**Comment by rtfb** [comment]
In general, this sounds overly tailored to HTML to my taste. With extensions like this markdown is turning into a dialect of HTML that has all the weaknesses of markdown underspecification and none of the strengths.

I'd like a more general solution. I was thinking about automatically generating some attributes and adding them to the elements. For example, every image could get a numbered attribute, `image-1`, `image-2`, etc., which could be rendered to classes in HTML case and to something else in other renderers. And best of all, the source document would remain untouched.

**Comment by lebarde** [comment]
The point is that I really want to write in Markdown without putting HTML. Markdown is not a specification to generate only HTML. It is a specification to generate HTML, PDFs, LaTeX, Org-mode, plain-text, ePUBs and so on. Pandoc is good at that, for instance.

Texts I write could be reused later on other supports, and for that I do need to write Markdown, or another sustainable plain-text specification. I chose markdown for its accessibility, but I am aware of the limitations of its first specification.

On the other hand, what I can do when I switch from one support to another is doing a targeted text review, like adding classes for a specific markdown variant (I would prefer not to do so, but this is doable).

Do you think it is possible to do that?

For now, **[I have to do this](https://discuss.gohugo.io/t/script-use-advanced-markdown-features-images-etc/6812)**, but it is a non-efficient hack to fit my needs.

**Comment by rdwatters** [comment]
> For example, every image could get a numbered attribute, image-1, image-2, etc., which could be rendered to classes in HTML case and to something else in other renderers. 

I like the line of reasoning for other output types, but it adds little benefit as a CSS class since the numeric, auto-generated classes would be superfluous to pseudoselectors, which already take care of this at the stylesheet level without unnecessary, non-semantic, markup (e.g., `:nth-child(2)`,`:first-child`, etc will take care of this). 

> And best of all, the source document would remain untouched.

Couldn't agree more, but perhaps a config option in BF for users who want to add this extra markup to their markdown? Then they have nobody but themselves to blame when they're managing bloated, quasi-HTML content files. Clearly this feature isn't my preference&mdash;I'd rather use the obvious solution from @shurcooL&mdash;but it does come up on the Hugo forums from time to time. 

The second option offered by @lebarde&mdash;although I'm not thrilled about the syntax&mdash; is the most flexible for those who want to add attributes in addition to `class` or `id` (e.g., `data`). I think there is a better syntactical approach, but that seems like a cart-before-horse suggestion.

I'm also throwing this out there with very little skin in the game, so take these suggestions accordingly :smile: Thanks!
